### ISSUE TITLE
(3.0.0-3.5.1) ParallelCluster CLI raises exception “module 'flask.json' has no attribute 'JSONEncoder'”

### ISSUE THREAD

**Issue by hanwen-pcluste** [aws/aws-parallelcluster#5244]
## Bug description

Flask is a Python dependency of ParallelCluster. Due to a recent change in flask, when flask > 2.2.4 is installed, the ParallelCluster CLI will raise an exception when executing any command. If affected by this issue, the pcluster command will raise an exception with the following message:
```
AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
```

## Mitigation

[(3.0.0-3.5.1) ParallelCluster CLI raises exception “module 'flask.json' has no attribute 'JSONEncoder'”](https://github.com/aws/aws-parallelcluster/wiki/(3.0.0-3.5.1)-ParallelCluster-CLI-raises-exception-%E2%80%9Cmodule-'flask.json'-has-no-attribute-'JSONEncoder'%E2%80%9D)

**Comment by Nekenlight** [comment]
Thanks

**Comment by hahnd** [comment]
Additional details on how to fix and a workaround:

### Download the Python 3.9 docker image

**> docker pull python:3.9**
```
3.9: Pulling from library/python
d677c78be691: Pull complete
73587d290df1: Pull complete
bd49b43770af: Pull complete
e9a96e84daab: Pull complete
1fae52722eac: Pull complete
23760ffdb04d: Pull complete
ad7cfc4d8f1e: Pull complete
690d2ae43687: Pull complete
Digest: sha256:6ea9dafc96d7914c5c1d199f1f0195c4e05cf017b10666ca84cb7ce8e2699d51
Status: Downloaded newer image for python:3.9
docker.io/library/python:3.9
```

### Run a Python 3.9 container and install aws-parallelcluster==3.2.1.  This installs flask-2.3.2 as a dependency
**> docker run -it python:3.9 /bin/bash**
```
root@141e070eaae6:/# python3 --version
Python 3.9.16
root@141e070eaae6:/# python3 -m pip install aws-parallelcluster==3.2.1
Collecting aws-parallelcluster==3.2.1
  Downloading aws_parallelcluster-3.2.1-py3-none-any.whl (424 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 425.0/425.0 KB 6.5 MB/s eta 0:00:00
...
...
...
Collecting flask~=2.0
  Downloading Flask-2.3.2-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.9/96.9 KB 12.8 MB/s eta 0:00:00
...
...
...
Installing collected packages: publication, zipp, urllib3, typing-extensions, typeguard, tabulate, six, PyYAML, pyrsistent, packaging, MarkupSafe, jmespath, itsdangerous, inflection, idna, exceptiongroup, click, charset-normalizer, certifi, blinker, attrs, werkzeug, requests, python-dateutil, marshmallow, jsonschema, jinja2, importlib-resources, importlib-metadata, clickclick, cattrs, jsii, flask, botocore, s3transfer, constructs, connexion, aws-cdk.region-info, aws-cdk.cloud-assembly-schema, boto3, aws-cdk.cx-api, aws-cdk.core, aws-cdk.aws-signer, aws-cdk.aws-sam, aws-cdk.aws-imagebuilder, aws-cdk.aws-iam, aws-cdk.aws-codestarnotifications, aws-cdk.aws-acmpca, aws-cdk.assets, aws-cdk.aws-kms, aws-cdk.aws-events, aws-cdk.aws-codeguruprofiler, aws-cdk.aws-cloudwatch, aws-cdk.aws-autoscaling-common, aws-cdk.aws-ssm, aws-cdk.aws-sqs, aws-cdk.aws-s3, aws-cdk.aws-ecr, aws-cdk.aws-applicationautoscaling, aws-cdk.aws-sns, aws-cdk.aws-s3-assets, aws-cdk.aws-ecr-assets, aws-cdk.aws-logs, aws-cdk.aws-codecommit, aws-cdk.aws-stepfunctions, aws-cdk.aws-kinesis, aws-cdk.aws-ec2, aws-cdk.aws-fsx, aws-cdk.aws-elasticloadbalancing, aws-cdk.aws-efs, aws-cdk.aws-lambda, aws-cdk.aws-sns-subscriptions, aws-cdk.aws-secretsmanager, aws-cdk.aws-cloudformation, aws-cdk.custom-resources, aws-cdk.aws-codebuild, aws-cdk.aws-route53, aws-cdk.aws-globalaccelerator, aws-cdk.aws-dynamodb, aws-cdk.aws-certificatemanager, aws-cdk.aws-elasticloadbalancingv2, aws-cdk.aws-cognito, aws-cdk.aws-cloudfront, aws-cdk.aws-servicediscovery, aws-cdk.aws-autoscaling, aws-cdk.aws-apigateway, aws-cdk.aws-route53-targets, aws-cdk.aws-autoscaling-hooktargets, aws-cdk.aws-ecs, aws-cdk.aws-batch, aws-parallelcluster
Successfully installed MarkupSafe-2.1.2 PyYAML-5.4.1 attrs-23.1.0 aws-cdk.assets-1.201.0 aws-cdk.aws-acmpca-1.201.0 aws-cdk.aws-apigateway-1.201.0 aws-cdk.aws-applicationautoscaling-1.201.0 aws-cdk.aws-autoscaling-1.201.0 aws-cdk.aws-autoscaling-common-1.201.0 aws-cdk.aws-autoscaling-hooktargets-1.201.0 aws-cdk.aws-batch-1.201.0 aws-cdk.aws-certificatemanager-1.201.0 aws-cdk.aws-cloudformation-1.201.0 aws-cdk.aws-cloudfront-1.201.0 aws-cdk.aws-cloudwatch-1.201.0 aws-cdk.aws-codebuild-1.201.0 aws-cdk.aws-codecommit-1.201.0 aws-cdk.aws-codeguruprofiler-1.201.0 aws-cdk.aws-codestarnotifications-1.201.0 aws-cdk.aws-cognito-1.201.0 aws-cdk.aws-dynamodb-1.201.0 aws-cdk.aws-ec2-1.201.0 aws-cdk.aws-ecr-1.201.0 aws-cdk.aws-ecr-assets-1.201.0 aws-cdk.aws-ecs-1.201.0 aws-cdk.aws-efs-1.201.0 aws-cdk.aws-elasticloadbalancing-1.201.0 aws-cdk.aws-elasticloadbalancingv2-1.201.0 aws-cdk.aws-events-1.201.0 aws-cdk.aws-fsx-1.201.0 aws-cdk.aws-globalaccelerator-1.201.0 aws-cdk.aws-iam-1.201.0 aws-cdk.aws-imagebuilder-1.201.0 aws-cdk.aws-kinesis-1.201.0 aws-cdk.aws-kms-1.201.0 aws-cdk.aws-lambda-1.201.0 aws-cdk.aws-logs-1.201.0 aws-cdk.aws-route53-1.201.0 aws-cdk.aws-route53-targets-1.201.0 aws-cdk.aws-s3-1.201.0 aws-cdk.aws-s3-assets-1.201.0 aws-cdk.aws-sam-1.201.0 aws-cdk.aws-secretsmanager-1.201.0 aws-cdk.aws-servicediscovery-1.201.0 aws-cdk.aws-signer-1.201.0 aws-cdk.aws-sns-1.201.0 aws-cdk.aws-sns-subscriptions-1.201.0 aws-cdk.aws-sqs-1.201.0 aws-cdk.aws-ssm-1.201.0 aws-cdk.aws-stepfunctions-1.201.0 aws-cdk.cloud-assembly-schema-1.201.0 aws-cdk.core-1.201.0 aws-cdk.custom-resources-1.201.0 aws-cdk.cx-api-1.201.0 aws-cdk.region-info-1.201.0 aws-parallelcluster-3.2.1 blinker-1.6.2 boto3-1.26.133 botocore-1.29.133 cattrs-22.2.0 certifi-2023.5.7 charset-normalizer-3.1.0 click-8.1.3 clickclick-20.10.2 connexion-2.13.1 constructs-3.4.294 exceptiongroup-1.1.1 flask-2.3.2 idna-3.4 importlib-metadata-6.6.0 importlib-resources-5.12.0 inflection-0.5.1 itsdangerous-2.1.2 jinja2-3.1.2 jmespath-0.10.0 jsii-1.81.0 jsonschema-4.17.3 marshmallow-3.19.0 packaging-23.1 publication-0.0.3 pyrsistent-0.19.3 python-dateutil-2.8.2 requests-2.30.0 s3transfer-0.6.1 six-1.16.0 tabulate-0.9.0 typeguard-2.13.3 typing-extensions-4.5.0 urllib3-1.26.15 werkzeug-2.3.4 zipp-3.15.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
WARNING: You are using pip version 22.0.4; however, version 23.1.2 is available.
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.
```

### Run Python and try to import run from pcluster.cli.entrypoint, has an error:
**root@141e070eaae6:/# python3**
```
Python 3.9.16 (main, May  4 2023, 02:16:04)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
```

```
>>> from pcluster.cli.entrypoint import run as pcluster
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/pcluster/cli/entrypoint.py", line 27, in <module>
    import pcluster.api.controllers.cluster_compute_fleet_controller
  File "/usr/local/lib/python3.9/site-packages/pcluster/api/controllers/cluster_compute_fleet_controller.py", line 11, in <module>
    from pcluster.api.controllers.common import configure_aws_region, convert_errors, validate_cluster
  File "/usr/local/lib/python3.9/site-packages/pcluster/api/controllers/common.py", line 22, in <module>
    from pcluster.api.errors import (
  File "/usr/local/lib/python3.9/site-packages/pcluster/api/errors.py", line 9, in <module>
    from connexion import ProblemException
  File "/usr/local/lib/python3.9/site-packages/connexion/__init__.py", line 32, in <module>
    from .apps.flask_app import FlaskApp
  File "/usr/local/lib/python3.9/site-packages/connexion/apps/flask_app.py", line 151, in <module>
    class FlaskJSONEncoder(json.JSONEncoder):
AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
```

### Downgrade flask to v2.2.3 and everything is ok:
**root@141e070eaae6:/# python3 -m pip install flask==2.2.3**
```
Collecting flask==2.2.3
  Downloading Flask-2.2.3-py3-none-any.whl (101 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.8/101.8 KB 3.8 MB/s eta 0:00:00
Requirement already satisfied: Jinja2>=3.0 in /usr/local/lib/python3.9/site-packages (from flask==2.2.3) (3.1.2)
Requirement already satisfied: Werkzeug>=2.2.2 in /usr/local/lib/python3.9/site-packages (from flask==2.2.3) (2.3.4)
Requirement already satisfied: click>=8.0 in /usr/local/lib/python3.9/site-packages (from flask==2.2.3) (8.1.3)
Requirement already satisfied: itsdangerous>=2.0 in /usr/local/lib/python3.9/site-packages (from flask==2.2.3) (2.1.2)
Requirement already satisfied: importlib-metadata>=3.6.0 in /usr/local/lib/python3.9/site-packages (from flask==2.2.3) (6.6.0)
Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.9/site-packages (from importlib-metadata>=3.6.0->flask==2.2.3) (3.15.0)
Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.9/site-packages (from Jinja2>=3.0->flask==2.2.3) (2.1.2)
Installing collected packages: flask
  Attempting uninstall: flask
    Found existing installation: Flask 2.3.2
    Uninstalling Flask-2.3.2:
      Successfully uninstalled Flask-2.3.2
Successfully installed flask-2.2.3
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
WARNING: You are using pip version 22.0.4; however, version 23.1.2 is available.
You should consider upgrading via the '/usr/local/bin/python3 -m pip install --upgrade pip' command.
```

### Now the module can be imported without error
**root@141e070eaae6:/# python3**
```
Python 3.9.16 (main, May  4 2023, 02:16:04)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pcluster.cli.entrypoint import run as pcluster
>>>
```

**Comment by steevmi1** [comment]
It looks like changing the flask version doesn't work for parallelcluster installed into a virtual environment (on OS X 12.6.5):

```
steevesm on osxlaptop@10.122.12.2 at 2023-05-18 08:08 in ~
via  pyenv (3.5.1-python-3.11.3) on  hpc (us-east-1) took 427ms ✖ pip list format==freeze
Package                             Version
----------------------------------- --------
attrs                               23.1.0
aws-cdk.assets                      1.201.0
aws-cdk.aws-acmpca                  1.201.0
aws-cdk.aws-apigateway              1.201.0
aws-cdk.aws-applicationautoscaling  1.201.0
aws-cdk.aws-autoscaling             1.201.0
aws-cdk.aws-autoscaling-common      1.201.0
aws-cdk.aws-autoscaling-hooktargets 1.201.0
aws-cdk.aws-batch                   1.201.0
aws-cdk.aws-certificatemanager      1.201.0
aws-cdk.aws-cloudformation          1.201.0
aws-cdk.aws-cloudfront              1.201.0
aws-cdk.aws-cloudwatch              1.201.0
aws-cdk.aws-codebuild               1.201.0
aws-cdk.aws-codecommit              1.201.0
aws-cdk.aws-codeguruprofiler        1.201.0
aws-cdk.aws-codestarnotifications   1.201.0
aws-cdk.aws-cognito                 1.201.0
aws-cdk.aws-dynamodb                1.201.0
aws-cdk.aws-ec2                     1.201.0
aws-cdk.aws-ecr                     1.201.0
aws-cdk.aws-ecr-assets              1.201.0
aws-cdk.aws-ecs                     1.201.0
aws-cdk.aws-efs                     1.201.0
aws-cdk.aws-elasticloadbalancing    1.201.0
aws-cdk.aws-elasticloadbalancingv2  1.201.0
aws-cdk.aws-events                  1.201.0
aws-cdk.aws-fsx                     1.201.0
aws-cdk.aws-globalaccelerator       1.201.0
aws-cdk.aws-iam                     1.201.0
aws-cdk.aws-imagebuilder            1.201.0
aws-cdk.aws-kinesis                 1.201.0
aws-cdk.aws-kms                     1.201.0
aws-cdk.aws-lambda                  1.201.0
aws-cdk.aws-logs                    1.201.0
aws-cdk.aws-route53                 1.201.0
aws-cdk.aws-route53-targets         1.201.0
aws-cdk.aws-s3                      1.201.0
aws-cdk.aws-s3-assets               1.201.0
aws-cdk.aws-sam                     1.201.0
aws-cdk.aws-secretsmanager          1.201.0
aws-cdk.aws-servicediscovery        1.201.0
aws-cdk.aws-signer                  1.201.0
aws-cdk.aws-sns                     1.201.0
aws-cdk.aws-sns-subscriptions       1.201.0
aws-cdk.aws-sqs                     1.201.0
aws-cdk.aws-ssm                     1.201.0
aws-cdk.aws-stepfunctions           1.201.0
aws-cdk.cloud-assembly-schema       1.201.0
aws-cdk.core                        1.201.0
aws-cdk.custom-resources            1.201.0
aws-cdk.cx-api                      1.201.0
aws-cdk.region-info                 1.201.0
aws-parallelcluster                 3.5.1
boto3                               1.26.135
botocore                            1.29.135
cattrs                              22.2.0
certifi                             2023.5.7
charset-normalizer                  3.1.0
click                               8.1.3
clickclick                          20.10.2
connexion                           2.13.1
constructs                          3.4.301
Flask                               2.2.5
idna                                3.4
importlib-resources                 5.12.0
inflection                          0.5.1
itsdangerous                        2.1.2
Jinja2                              3.1.2
jmespath                            0.10.0
jsii                                1.81.0
jsonschema                          4.17.3
MarkupSafe                          2.1.2
marshmallow                         3.19.0
packaging                           23.1
pip                                 23.1.2
publication                         0.0.3
pyrsistent                          0.19.3
python-dateutil                     2.8.2
PyYAML                              5.4.1
requests                            2.30.0
s3transfer                          0.6.1
setuptools                          67.7.2
six                                 1.16.0
tabulate                            0.8.10
typeguard                           2.13.3
typing_extensions                   4.5.0
urllib3                             1.26.15
Werkzeug                            2.3.4
steevesm on osxlaptop@10.122.12.2 at 2023-05-18 08:09 in ~
via  pyenv (3.5.1-python-3.11.3) on  hpc (us-east-1) took 470ms ➜ type -a pcluster
pcluster is /Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/bin/pcluster
steevesm on osxlaptop@10.122.12.2 at 2023-05-18 08:10 in ~
via  pyenv (3.5.1-python-3.11.3) on  hpc (us-east-1) took 7ms ➜ pcluster --help
Traceback (most recent call last):
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/bin/pcluster", line 5, in <module>
    from pcluster.cli.entrypoint import main
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/cli/entrypoint.py", line 27, in <module>
    import pcluster.api.controllers.cluster_compute_fleet_controller
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/api/controllers/cluster_compute_fleet_controller.py", line 11, in <module>
    from pcluster.api.controllers.common import configure_aws_region, convert_errors, validate_cluster
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/api/controllers/common.py", line 33, in <module>
    from pcluster.models.cluster import Cluster
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/models/cluster.py", line 35, in <module>
    from pcluster.config.cluster_config import BaseClusterConfig, SchedulerPluginScheduling, SlurmScheduling, Tag
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/config/cluster_config.py", line 111, in <module>
    from pcluster.validators.directory_service_validators import (
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/pcluster/validators/directory_service_validators.py", line 15, in <module>
    from aws_cdk.core import Arn
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/aws_cdk/core/__init__.py", line 988, in <module>
    from ._jsii import *
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/aws_cdk/core/_jsii/__init__.py", line 13, in <module>
    import aws_cdk.cloud_assembly_schema._jsii
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/aws_cdk/cloud_assembly_schema/__init__.py", line 77, in <module>
    from ._jsii import *
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/aws_cdk/cloud_assembly_schema/_jsii/__init__.py", line 13, in <module>
    __jsii_assembly__ = jsii.JSIIAssembly.load(
                        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_runtime.py", line 55, in load
    _kernel.load(assembly.name, assembly.version, os.fspath(assembly_path))
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/__init__.py", line 299, in load
    self.provider.load(LoadRequest(name=name, version=version, tarball=tarball))
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/providers/process.py", line 352, in load
    return self._process.send(request, LoadResponse)
           ^^^^^^^^^^^^^
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_utils.py", line 23, in wrapped
    stored.append(fgetter(self))
                  ^^^^^^^^^^^^^
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/providers/process.py", line 347, in _process
    process.start()
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/providers/process.py", line 260, in start
    self._process = subprocess.Popen(
                    ^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 1024, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 1917, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'node'
Exception ignored in: <function _NodeProcess.__del__ at 0x107a53600>
Traceback (most recent call last):
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/providers/process.py", line 228, in __del__
    self.stop()
  File "/Users/steevesm/.local/pcluster/3.5.1-python-3.11.3/lib/python3.11/site-packages/jsii/_kernel/providers/process.py", line 291, in stop
    assert self._process.stdin is not None
           ^^^^^^^^^^^^^
AttributeError: '_NodeProcess' object has no attribute '_process'
```

**Comment by hahnd** [comment]
@steevmi1 Looks like you have a different error.  pcluster 3.x requires NodeJS.  Make sure you have [NodeJS installed](https://github.com/nvm-sh/nvm)  and in your path.

**Comment by enrico-usai** [comment]
Closed since we released 3.7.0 with an updated pin of the flask library.
